### ISSUE TITLE
UNetMotionModel.dtype is really expensive to call, is it possible to cache it during inference?

### ISSUE THREAD

**Issue by xiang9156** [huggingface/diffusers#9520]
**What API design would you like to have changed or added to the library? Why?**
we are using class UNetMotionModel(ModelMixin, ConfigMixin, UNet2DConditionLoadersMixin, PeftAdapterMixin)
and its `forward()` implementation is calling self.dtype, which is very expensive
![image](https://github.com/user-attachments/assets/cb840057-ccf7-46ed-847d-2c8aef292fe9)
from my profiling trace result, calling self.dtype takes 6-10ms each time.
can we somehow cache it to save time?
![image](https://github.com/user-attachments/assets/b5ef3c1e-ee9f-4f02-922e-854ebe269568)

I took a look at ModelMixin.dtype() property function,  it get all parameters of the model into tuple to check only first parameter's dtype, i don't thinkmake sense to do this everytime. right?
![image](https://github.com/user-attachments/assets/b74a8c31-0b4e-44cb-ab09-e3f7c5559dad)

**What use case would this enable or better enable? Can you give us a code example?**
We are using this model to do video generation, so the inference is running repeatedly. Is it easy to optimize this ~10ms latency?
Thanks!

**Comment by yiyixuxu** [comment]
thanks! yes!

**Comment by yiyixuxu** [comment]
I think we should optimize `get_parameter_dtype` a bit https://github.com/huggingface/diffusers/blob/28f9d84549c0b1d24ef00d69a4c723f3a11cffb6/src/diffusers/models/modeling_utils.py#L94

**Comment by a-r-r-o-w** [comment]
Thank you for actively helping improve performance bottlenecks! I will take a look at the trace too, sometime soon, to confirm this.

> I think we should optimize get_parameter_dtype a bit

Since we assume that the underlying object that we're checking is a torch.nn.Module, do you think having `next(parameter.parameters()).dtype` would be better here? That way, we wouldn't have to create a tuple and can use the underlying iterators that the torch module already provides. No idea if it's an optimization though, and the part after StopIteration may still have to remain. What do you have in mind?

**Comment by yiyixuxu** [comment]
@a-r-r-o-w yep yep that's exactly what I thought, tested it out and it is a lot faster https://github.com/huggingface/diffusers/pull/9526

it returns once it found the dtype so would not normally need to run the code path after StopIteration (not sure when that code path is needed actually lol, notes says it is for pytorch 1.5)

**Comment by a-r-r-o-w** [comment]
not stale. cc @DN6
